### PR TITLE
WIP: Microboluses

### PR DESCRIPTION
**Pull request by ivalkou**
Microboluses use internal Loop system to suggest boluses when basal change is not enough, but it forces a part of that dosage in the beginning of every loop. Be sure that you settings are correct because this system is more sensitive to mistakes than the original one. Also, microboluses are a lot more sensitive to BG sensor noise, so definitely keep that in mind

For microbolusing to work, be sure that the allowed basal increase is as minimal as possible (difference between basal delivery limit and current basal rate). The less it is, the harsher microboluses get

The part of the dosage is also configurable. By default it is set to 30%. That means that if Loops suggests a bolus of 1 unit of insulin at the beginning of the loop, 0.3 units will be implemented automatically. You can increase and decrease this setting to make microboluses more aggressive.

Maximal microbolus amount is set in basal minutes. This value determines the amount of basal insulin from your current profile for the given number of minutes (not TBR). For example, if the basal rate is 1 unit/hour, the 60 basal minutes amount is 1 unit. Basal minutes are also used in other artificial pancreas system (like OpenAPS or AndroidAPS), that's why they are used in Microbolus configuration. You can also choose not to limit the microbolus amount.

Microbolusing limits can be set separately to the situations when COB is 0 and when it's more that 0

Microboluses safe mode can be set to one of three options:
- _Disabled_ - no additional limits
- _Limited_ - microboluses are limited to 30 basal minutes (additionaly to your microboluses limit) when the BG is predicted to fall in 15 minutes
- _Enabled_ - microboluses will not be enacted when BG is predicted to fall

Microboluses limits

- Current BG is lower than a suspend treshold
Microboluses are off

- Current BG is below the target range
Maximal microbolus amount is limited to 30 basal minutes

- Predicted BG at any moment is below the suspension treshold
Microboluses are off

- Eventual predicted BG is below the suspension treshold
Microboluses are off

- Eventual predicted BG is below the target range
Microboluses are off

In any other case microboluses are limited by your settings

### CODE REVIEW

**Comment by ivalkou** [comment]
Known issues:
- A52 risk (Medtrinic pumps)

_Updating..._

**Comment by loyalchallenger** [comment]
Will there be an option to not use this feature?

**Comment by ps2** [comment]
There are likely significant changes for this feature. It's in the testing phase. It will be available as a testing feature for those who want to try it out and are willing to be testers. If, after a while, it seems to be a good thing for most people, it would probably land in dev/master as an option, and only if it seemed prevailingly a better choice for users would it become the default, assuming other risks were sufficiently mitigated.

**Comment by jbwittmer** [comment]
I'm sorry I've not been keeping up on the progress in dev, but since you can't query the reservoir volume for the Omnipod, what is the safety mechanism to avoid multiple boluses? I recall this being an concern on the OpenAPS side that was addressed in part through the mechanism of querying reservoir volume before/after a bolus command to assure that it was/wasn't successful.  Or feel free to point me elsewhere if I should read up on the developments.  Thanks.

**Comment by ps2** [comment]
There is no separate user interface for pods, so Loop should know about all boluses on Omnipod. Still, there needs to be an additional mitigation in the Loop code so that the user doesn't issue a bolus while the automatic bolus has already delivered behind the bolus entry screen.

**Comment by ps2** [comment]
This is a very experimental branch. Anyone trying it at this point should do so either disconnected to play with it, or do so only after fully understanding how it works, and what risks there might be. Bolusing while a bolus has been triggered in the background is one of them.

**Comment by bartsluimer** [comment]
Assume this is the same as the version from Ivalkou Lookworkspace, can confirm that it is working pretty well. We use it as way to automatic bolus when values are to far off. 

It is important to lower your max basal to for example 2x your basal. Otherwise it will not work. 
No issues even when we had sensor change and we sometimes have got 3 moll/L difference between real and sensor value.

We see an improvement with using Micro Bolussing.

**Comment by Kdisimone** [comment]
@ivalkou tiny tiny nit...can you add a commit to remove your Dev Team from the targets' signing (I think technically you'd have 5 to set to "none"...the extra being the release signing in WatchApp Extension target). I'd be happy to do a pull request to your branch if that makes it easier for you. As more people try this out, I'm just hoping to avoid the "what's the team" questions in Looped ;) Thanks

File: Loop/Views/MicrobolusView.swift [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+                        Text(viewModel.unit.localizedShortUnitString)
+                    }
+                }
+            }
+
+            Section(header: Text("Other Options").font(.headline), footer:
+                Text("This is the minimum microbolus size in units that will be delivered. Only if the microbolus calculated is equal to or greater than this number of units will a bolus be delivered.")
+            ) {
+                Toggle (isOn: $viewModel.openBolusScreen) {
+                    Text("Open Bolus screen after Carbs")
```
[the remark this answers is not in the document]
Comment (ivalkou) [reply Loop/Views/MicrobolusView.swift]: Let's listen to the suggestions of others

**Comment by ddaniels1** [comment]
@ivalkou thanks for this branch. We have been using it for the last few days with our 4 year old and working great. One thing I will throw out for your consideration is whether it would be possible to have separate partial application settings with and without cob. In the case with cob This would help us deal with a late re-rise that happens with slow GI foods like pasta several hours later. We set the pizza absorption time, but with high temping it’s not enough. Personally I would set a higher partial applicantion with cob to deal with this, but would like to have a lower % for instanced with no cob so loop doesn’t overdo it.

**Comment by ps2** [comment]
@ddaniels1 Adjusting partial application is probably the wrong way to deal with carbs.  If loop is mishandling carbs for you, then something about its carb model needs to change, or how you use the carb model.

**Comment by Nightfoxy** [comment]
@ddaniels1 are you seeing better results with slow food using this branch? We have the same problem. I don't think I have ever had loop keep her in range in the last half of slow foods. Even accounting for fat /protein in the initial entry, trying longer and shorter absorption, I end up needing to double the initial entry halfway through. Best results seem to be a separate fat/protein entry 90 minutes to 2 hours after the first entry. But I can't quite figure out the calculation for it. 

I was hoping MB would be the solution if the problem was that the temp basal changes are just too slow.

**Comment by ddaniels1** [comment]
@ps2, loops ability to work well depends on a significant amount of insulin coming from basal. For kids under 6, often 3/4 of their insulin comes from bolus. I believe this Amy be because per lb, they eat significantly more carbs than an adult, but that lower body weight requires a proportionally Lowe basal. Olivia was diagnosed when she was 1, she is now 4 and we are slowly seeing a transition to more basal, and therefore more effectiveness from the loop model. This is very common both in taking to other parents and our Endo. It has been my feeling that because of this their are some changes needed to accommodate this. I can’t say we know the answer and in reply to @nightfoxy, microboluses have helped, but we often add carbs 2-3 hours later, or recently i came up with a pasta override that runs for 8 hours after a pasta meal at 150% normal. This also seems to help.

**Comment by ddaniels1** [comment]
We have noticed I think since the update running microboluses we are using significantly more power and can’t make it through the day without charging. May be coincidence and battery is going bad on phone but wanted to see if others have seen similar issues? I tried rebooting and the metrics show the majority of power being used by loop.

**Comment by louis4052** [comment]
@ddaniels1 I have noticed the same with my phone battery since running microboluses .

**Comment by JELCRAWFORD** [comment]
FWIW - we have not noticed and difference in phone battery life since starting Ivan’s MB branch. using an older iPhone 7 with a battery replaced about 8 months ago.

**Comment by ivalkou** [comment]
All further development of this option is underway at https://github.com/ivalkou/LoopWorkspace. News and discussions in Telegram https://t.me/LoopAPSEng

**Pull request closed without merging**